src/Featurizers/BackwardFillImputerFeaturizer.h

```diff
@@ -0,0 +1,164 @@
+#include "../Traits.h"
+
+#include "Components/InferenceOnlyFeaturizerImpl.h"
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         BackwardFillImputerTransformer
+///  \brief         Imputer that populates previous incoming null values with
+///                 the current value.
+///
+template <typename T>
+class BackwardFillImputerTransformer :
+    public Transformer<
+        typename Traits<T>::nullable_type,
+        T
+    > {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    static_assert(Traits<T>::IsNullableType == false || Traits<T>::IsNativeNullableType, "'T' must not be a nullable type");
+
+    using BaseType =
+        Transformer<
+            typename Traits<T>::nullable_type,
+            T
+        >;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    BackwardFillImputerTransformer(void);
+    BackwardFillImputerTransformer(Archive &ar);
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(BackwardFillImputerTransformer);
+
+    bool has_pending(void) const;
+    BackwardFillImputerTransformer & reset(void);
+
+    void save(Archive &ar) const override;
+
+    bool operator==(BackwardFillImputerTransformer const &other) const;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    std::uint32_t                           _numPending;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the definition and declaration are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        // ----------------------------------------------------------------------
+        using TheseTraits                   = Traits<typename BaseType::InputType>;
+        // ----------------------------------------------------------------------
+
+        if(TheseTraits::IsNull(input)) {
+            ++_numPending;
+            return;
+        }
+
+        T const &                           value(TheseTraits::GetNullableValue(input));
+        std::uint32_t                       toWrite(_numPending + 1);
+
+        while(toWrite--)
+            callback(value);
+
+        _numPending = 0;
+    }
+
+    void flush_impl(typename BaseType::CallbackFunction const &) override {

```

Comment:
And where would this method be used? I didn't find an usage of this.

Reply:
This is called from a parent class

---
In reply to: [377925921](https://github.com/microsoft/FeaturizersLibrary/pull/139#discussion_r377925921) [](ancestors = 377925921)